equinox-shipping/Domain/FinalizationTransaction.fs

```diff
@@ -0,0 +1,100 @@
+        | Completed
+        interface TypeShape.UnionContract.IUnionContract
+
+    let codec = FsCodec.NewtonsoftJson.Codec.Create<Event>()
+
+type Action =
+    | AssignShipments   of containerId : string<containerId> * shipmentIds : string<shipmentId>[]
+    | FinalizeContainer of containerId : string<containerId>
+    // Reverts the assignment of the container for the shipments provided.
+    | RevertAssignment  of shipmentIds : string<shipmentId>[]
+    | Finish            of success : bool
+
+module Fold =
+
+    type State = FinalizationTransactionState
+
+    let initial: State = { container = None; state = Initial }
+
+    let evolve (state : State) (event : Events.Event): State =
+        let tag (ids: string[]) = ids |> Array.map UMX.tag
+
+        match state, event with
+        | _, Events.FinalizationRequested event ->
+            { container = Some (UMX.tag event.containerId); state = Running (Assigning (tag event.shipmentIds)) }
+
+        | _, Events.AssignmentCompleted event ->
+            { state with state = Running (Assigned (tag event.shipmentIds)) }
+
+        | _, Events.RevertRequested event ->
+            { state with state = Running (Reverting (tag event.shipmentIds)) }
+
+        | { state = Running (Assigned _) }, Events.Completed ->
+            { state with state = Completed true }
+
+        | _, Events.Completed ->
+            { state with state = Completed false }
+
+    let nextAction (state: State): Action =
+        match state with
+        | { container = Some cId; state = Running (Assigning shipmentIds) } -> Action.AssignShipments   (cId, shipmentIds)
+        | { container = Some cId; state = Running (Assigned  _) }           -> Action.FinalizeContainer cId
+        | { state = Running (Reverting shipmentIds) }                       -> Action.RevertAssignment  shipmentIds
+        | { state = Completed result }                                      -> Action.Finish result
+        | s -> failwith (sprintf "Cannot interpret state %A" s)
+
+    let fold: State -> Events.Event seq -> State =
+        Seq.fold evolve
+
+    let chooseValidTransition (event : Events.Event) (state : State) =
+        match state, event with
+        | { state = Initial },               Events.FinalizationRequested _
+        | { state = Running (Assigning _) }, Events.AssignmentCompleted _
+        | { state = Running (Assigning _) }, Events.RevertRequested _
+        | { state = Running (Assigned _)  }, Events.Completed
+        | { state = Running (Reverting _) }, Events.Completed -> Some event
+        | _ -> None
+
+// When there are no event to apply to the state, it pushes the transaction manager to
+// follow up on the next action from where it was.
+let decide (update : Events.Event option) (state : Fold.State) : Action * Events.Event list =
+    let events =
+        match update with
+        | Some e -> Fold.chooseValidTransition e state |> Option.toList
```

Comment:
isnt there an Option.choose? (if there isnt, make it a bool predicate and do `Option.filter`)